### PR TITLE
[ISSUE #5537]🚀Implement ControllerApplyBrokerId Request Handler

### LINKED ISSUE #5537: [Feature🚀] Implement ControllerApplyBrokerId Request Handler
# Feature Request: Implement ControllerApplyBrokerId Request Handler

## Feature Description

Implement the `handle_apply_broker_id` method in `ControllerRequestProcessor` to handle requests for applying a specific broker ID. This feature allows brokers to request a particular broker ID, typically used during broker restart or migration scenarios.

## Problem/Motivation

Currently, the `handle_apply_broker_id` method is unimplemented. Applying for a specific broker ID is essential for:
- Broker restart with same ID (maintaining identity across restarts)
- Broker migration/replacement (new instance taking over old broker's ID)
- Disaster recovery scenarios (restoring from backup with known ID)
- Deterministic broker ID assignment (pre-planning cluster topology)
- Maintaining consistent routing and client connections

Without this implementation, brokers cannot retain their IDs across restarts, potentially causing:
- Disruption to client routing tables
- Loss of broker identity and history
- Complications during planned maintenance
- Issues with stateful operations tied to broker ID

### Location
- **File**: `rocketmq-controller/src/processor/controller_request_processor.rs`
- **Method**: `handle_apply_broker_id` (lines 446-459)
- **Request Code**: `RequestCode::ControllerApplyBrokerId`

### Implementation Steps

1. **Decode Request Header**:
   - Create/use `ApplyBrokerIdRequestHeader` in `rocketmq-remoting/protocol/header/controller/`
   - Header should contain:
     - `cluster_name`: The cluster name
     - `broker_name`: The broker group name
     - `applied_broker_id`: The specific broker ID requested
     - `register_check_code`: Optional verification code for security

2. **Validate Request**:
   - Check if requested broker ID is valid (within allowed range)
   - Check if ID is already in use by another broker
   - Verify broker has authority to claim this ID (via check code or other auth)
   - For ID 0 (master), may have additional requirements

3. **Apply Broker ID**:
   - Call `self.controller_manager.controller().apply_broker_id(&request_header).await`
   - This requires Raft consensus to reserve the ID
   - Logic:
     - If ID is free: Allocate to requesting broker
     - If ID was previously owned by same broker: Allow reclaim
     - If ID is in use: Reject request
     - Update ID allocation state

4. **Return Response**:
   - Return response indicating success or failure
   - On success: Confirm allocated broker ID
   - On failure: Explain why (ID in use, invalid, etc.)

### Example Implementation Pattern

```rust
async fn handle_apply_broker_id(
    &mut self,
    _channel: Channel,
    _ctx: ConnectionHandlerContext,
    request: &mut RemotingCommand,
) -> RocketMQResult<Option<RemotingCommand>> {
    use rocketmq_error::RocketMQError;
    use rocketmq_remoting::protocol::header::controller::apply_broker_id_request_header::ApplyBrokerIdRequestHeader;
    use rocketmq_remoting::protocol::header::controller::apply_broker_id_response_header::ApplyBrokerIdResponseHeader;

    // Decode request header
    let request_header = request
        .decode_command_custom_header::<ApplyBrokerIdRequestHeader>()
        .map_err(|e| {
            RocketMQError::request_header_error(format!(
                "Failed to decode ApplyBrokerIdRequestHeader: {:?}", e
            ))
        })?;

    // Validate requested broker ID
    if request_header.applied_broker_id < 0 {
        return Ok(Some(
            RemotingCommand::create_response_command_with_code_remark(
                ResponseCode::SystemError,
                "Invalid broker ID".to_string(),
            )
        ));
    }

    // Apply for specific broker ID via controller (requires consensus)
    self.controller_manager
        .controller()
        .apply_broker_id(&request_header)
        .await
}
```

## Alternatives

- Could always force brokers to get new ID on restart, but this breaks identity
- Could use external persistent storage for broker ID mapping
- Could use broker address as identifier instead of numeric ID

### Dependencies Required

1. **Request Header Type**:
   - Create `ApplyBrokerIdRequestHeader` in `rocketmq-remoting/src/protocol/header/controller/`
   - Fields:
     - `cluster_name: String`
     - `broker_name: String`
     - `applied_broker_id: u64` - requested broker ID
     - `register_check_code: Option<String>` - verification code

2. **Response Header Type**:
   - Create `ApplyBrokerIdResponseHeader` in `rocketmq-remoting/src/protocol/header/controller/`
   - Fields:
     - `broker_id: u64` - confirmed broker ID (may differ from request)
     - `result_code: String` - success/failure reason

3. **Controller Method**:
   - Implement `apply_broker_id(&self, header: &ApplyBrokerIdRequestHeader) -> RocketMQResult<Option<RemotingCommand>>`
   - Requires Raft consensus to reserve/allocate ID
   - Should check ID availability and ownership

4. **ID Ownership Tracking**:
   - Track which broker instance (by address) owns which ID
   - Allow reclaiming own ID after restart
   - Prevent stealing another broker's ID
   - Handle cleanup when broker is removed

### Related Code

- Related to `handle_get_next_broker_id` which auto-allocates next ID
- Related to `handle_register_broker` which uses the allocated ID
- Related to `handle_clean_broker_data` which releases IDs
- Should interact with broker ID allocation state machine

### Use Cases

1. **Broker restart**: Broker reclaims its previous ID
   - Broker stores its ID persistently
   - On restart, requests same ID via apply_broker_id
   - Controller verifies and approves

2. **Broker migration**: New instance takes over old broker's role
   - Old broker is stopped and cleaned up
   - New broker applies for the old broker's ID
   - Clients continue using same routing

3. **Disaster recovery**: Restoring broker from backup
   - Backup includes broker ID
   - Restored broker reclaims original ID

4. **Pre-planned topology**: Admin assigns specific IDs
   - Configuration specifies broker IDs
   - Brokers apply for their assigned IDs on startup

### Security Considerations

- **ID verification**: Use `register_check_code` to prevent unauthorized ID claiming
- **Ownership proof**: Broker should prove it previously owned the ID
- **Audit logging**: Log all ID application attempts
- **Rate limiting**: Prevent ID application spam
- **Admin override**: Allow admin to forcibly reassign IDs

### Implementation Notes

- This is a state-changing operation requiring Raft consensus
- Should validate ID is not currently in use by **active** broker
- May allow reclaiming ID from **offline** broker after timeout
- Should be idempotent (applying for same ID twice = same result)
- Need to distinguish between "ID never allocated" vs "ID was allocated but broker is offline"

### ID Application Logic

```
1. Check if ID is in valid range
2. Check if ID is currently in use by active broker
   - If yes and different broker: REJECT
   - If yes and same broker: APPROVE (idempotent)
3. Check if ID was previously allocated
   - If yes to this broker: APPROVE (reclaim)
   - If yes to different broker but offline: CONDITIONAL (timeout check)
   - If no: APPROVE (first allocation)
4. Reserve ID via Raft consensus
5. Return approval or rejection
```

### Edge Cases

- Multiple brokers request same ID concurrently: Raft ensures only one succeeds
- Broker offline but not cleaned: May need timeout before allowing reclaim
- Master ID (0) has special semantics: May require election or special approval
- ID exhaustion: Define maximum broker ID per group

### References

- Apache RocketMQ Java implementation: `ApplyBrokerIdRequestHeader` and processor
- Current file: [controller_request_processor.rs](rocketmq-controller/src/processor/controller_request_processor.rs#L446-L459)
- Related: [GetNextBrokerId issue](ControllerGetNextBrokerId.md), [RegisterBroker issue](ControllerRegisterBroker.md)

### PR DESCRIPTION
**Pull request by WaterWhisperer**
### Which Issue(s) This PR Fixes(Closes)

Fixes #5537 

### Brief Description

- Add `ControllerRequest::ApplyBrokerId` and `ControllerResponse::ApplyBrokerId` type
- Implement the `ApplyBrokerId` request processing logic in the state machine
- Implement the OpenRaft version of the `apply_broker_id` method
- Partially implements the raft-rs version of the `apply_broker_id` method
- Implement `handle_apply_broker_id` handler

### How Did You Test This Change?

Add some tests

## Summary by CodeRabbit

* **New Features**
  * Broker ID assignment flow implemented with validation, leadership checks, and enriched responses including cluster and broker metadata
* **Bug Fixes**
  * Clear error responses for invalid IDs, leadership/state issues, and ID conflicts; idempotent handling for repeated requests
* **Tests**
  * Added comprehensive tests covering success, idempotency, conflicts, edge cases, and header validation

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: rocketmq-controller/src/controller/open_raft_controller.rs [hunk 1]
```diff
@@ -177,9 +179,135 @@ impl Controller for OpenRaftController {
...
+
+        // Create the Raft request
+        let raft_request = crate::typ::ControllerRequest::ApplyBrokerId {
+            cluster_name: request.cluster_name.to_string(),
+            broker_name: request.broker_name.to_string(),
+            broker_addr: String::new(),
+            applied_broker_id: request.applied_broker_id as u64,
+            register_check_code: request.register_check_code.to_string(),
+        };
+
```
[the remark this answers is not in the document]
Comment (WaterWhisperer) [reply rocketmq-controller/src/controller/open_raft_controller.rs:246]: This meets the design of the Java version

File: rocketmq-controller/src/openraft/state_machine.rs [hunk 2]
```diff
@@ -193,6 +193,69 @@ impl StateMachine {
...
+                    register_check_code
+                );
+
+                ControllerResponse::ApplyBrokerId {
+                    success: true,
+                    error: None,
+                    cluster_name: cluster_name.clone(),
+                    broker_name: broker_name.clone(),
+                }
+            }
```
[the remark this answers is not in the document]
Comment (WaterWhisperer) [reply rocketmq-controller/src/openraft/state_machine.rs:257]: I think this involves overall architecture design

File: rocketmq-controller/src/controller/open_raft_controller.rs [hunk 3]
```diff
@@ -177,9 +179,135 @@ impl Controller for OpenRaftController {
...
+            }
+        }
+
+        tracing::info!(
+            "Processing ApplyBrokerId request: cluster={}, broker={}, broker_id={}, check_code={}",
+            request.cluster_name,
+            request.broker_name,
+            request.applied_broker_id,
+            request.register_check_code
+        );
```
[the remark this answers is not in the document]
Comment (WaterWhisperer) [reply rocketmq-controller/src/controller/open_raft_controller.rs:236]: I'll fix it
